File: pkg-py/tests/test_remote_orphanet.py
"""Orphanet remote mode: OLS lookup against ordo with an id-prefix rewrite."""

import pytest

import biogate
import biogate._remote as remote
from biogate import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def _found():
    return {"page": {"totalElements": 1}}


def test_url_rewrites_the_orpha_prefix_for_ols():
    from biogate._registry import get_source

    url = remote._ols_url(get_source("orphanet"), "ORPHA:558")
    assert url == (
        "https://www.ebi.ac.uk/ols4/api/ontologies/ordo/terms?obo_id=Orphanet:558"
    )


def test_existing_term_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _found()))
    res = biogate.check_id("ORPHA:558", source_db="orphanet", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "ORPHA:558"
    assert res.suggestion is None


def test_absent_term_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biogate.check_id("ORPHA:999999", source_db="orphanet", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _found()))
    res = biogate.check_id("orpha:558", source_db="orphanet", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "ORPHA:558"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biogate.check_id("ORPHA558", source_db="orphanet", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biogate.check_id("ORPHA:558", source_db="orphanet", how="remote")
    path = remote._remote_cache_path("ols", "ordo", "ORPHA:558")
    assert not path.is_file()

File: pkg-r/tests/testthat/test-remote-orphanet.R
# Offline tests for orphanet remote mode. Existence is checked against the ordo
# ontology in OLS, with the ORPHA prefix rewritten to Orphanet for the lookup;
# the biogate.remote_transport option replaces the network.

.stub_orphanet <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}
.orphanet_found <- function() {
  list(page = list(totalElements = 1))
}

test_that("the url rewrites the orpha prefix for ols", {
  source <- list(remote = list(ols_ontology = "ordo", obo_prefix = "Orphanet"))
  url <- .remote_resolvers$ols$url(source, "ORPHA:558")
  expect_identical(
    url,
    "https://www.ebi.ac.uk/ols4/api/ontologies/ordo/terms?obo_id=Orphanet:558"
  )
})

test_that("an existing term is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_orphanet(200, .orphanet_found())
  )
  res <- check_id("ORPHA:558", "orphanet", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "ORPHA:558")
  expect_true(is.na(res$suggestion))
})

test_that("an absent term is invalid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_orphanet(404))
  res <- check_id("ORPHA:999999", "orphanet", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_orphanet(200, .orphanet_found())
  )
  res <- check_id("orpha:558", "orphanet", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "ORPHA:558")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("ORPHA558", "orphanet", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_orphanet(500))
  expect_error(
    check_id("ORPHA:558", "orphanet", how = "remote"),
    class = "biogate_error_remote"
  )
  expect_false(file.exists(.remote_cache_path("ols", "ordo", "ORPHA:558")))
})
